src/structurizr/model/container.py

```diff
@@ -99,41 +98,68 @@ def __init__(
         super().__init__(**kwargs)
         self.parent = parent
         self.technology = technology
-        self.components = set(components)
+        self._components = set(components)
 
         self.tags.add(Tags.CONTAINER)
 
+    @property
+    def components(self) -> Iterable[Component]:
+        """Return read-only list of child components."""
+        return list(self._components)
+
     @classmethod
     def hydrate(
         cls,
         container_io: ContainerIO,
         software_system: "SoftwareSystem",
         model: "Model",
     ) -> "Container":
-        """"""
+        """Hydrate a new Container instance from its IO."""
         container = cls(
             **cls.hydrate_arguments(container_io),
             parent=software_system,
             technology=container_io.technology,
         )
+        model += container
 
         for component_io in container_io.components:
             component = Component.hydrate(component_io, container=container)
-            model.add_component(component, parent=container)
+            container += component
 
         return container
 
-    def add_component(self, component: Component = None, **kwargs) -> Component:
-        return self.get_model().add_component(
-            parent=self,
-            component=component,
-            **kwargs,
-        )
-
-    def add(self, component: Component) -> None:
-        self.components.add(component)
+    def add_component(self, **kwargs) -> Component:
+        """Add a new component to this container."""
+        component = Component(**kwargs)
+        self += component
+        return component
+
+    def __iadd__(self, component: Component) -> "Container":
+        """Add a newly constructed component to this container."""
+        # TODO: once we move past python 3.6 change to proper return type via
+        # __future__.annotations
+        if component in self.components:
+            return self
+
+        if self.get_component_with_name(component.name):
+            raise ValueError(
+                f"Component with name {component.name} already exists in {self}."
+            )
+
+        if component.parent is None:
+            component.parent = self
+        elif component.parent is not self:
+            raise ValueError(
+                f"Component with name {component.name} already has parent "
+                f"{component.parent}. Cannot add to {self}."
+            )
+        self._components.add(component)
+        model = self.get_model()
+        model += component
```

Comment:
I'd shorten this to:
```suggestion
        self.get_model()  += component
```
btw, there is also a property, so if you prefer you can write
```suggestion
        self.model += component
```

Reply:
Unfortunately that doesn't work as self.model is read-only, which is irritating.  I've added an `add()` method to Model instead which makes this code less ugly, but I'm not sure whether it'd confuse.

Reply:
Hmm, that is curious. Sorry I didn't expect that. I must admit that I don't like the addition of the `add` method that you made in https://github.com/Midnighter/structurizr-python/pull/33/commits/880affe39482ef893a8540ce5fbb51054cbd2bcd I'd rather that you revert those changes and then I'll merge.

Reply:
Yep, agreed.  Done.